From 8d2206ffde5cbc323ba8290a49f6be176dddd026 Mon Sep 17 00:00:00 2001
From: Gurpreet Ghai <gghai@codeaurora.org>
Date: Mon, 13 Aug 2018 18:58:45 +0530
Subject: [PATCH] BT: Update page-type and latency for Aptx Connections

-> Restrict 3Mbps packets for Aptx connections.
-> Update the sink latency value based on mode switch
   between LL and HQ for Aptx Adaptive

CRs-Fixed: 2298855

Change-Id: Iea929c8c15daafb6e61a0c9740eda7ae26b7286b
---
 bta/av/bta_av_aact.cc                 | 12 +++++++
 btif/include/btif_av.h                |  3 ++
 btif/src/btif_a2dp_audio_interface.cc |  5 +++
 btif/src/btif_av.cc                   | 63 +++++++++++++++++++++++++++++++++--
 stack/avdt/avdt_api.cc                | 16 +++++++++
 stack/include/avdt_api.h              |  2 ++
 6 files changed, 98 insertions(+), 3 deletions(-)

diff --git a/bta/av/bta_av_aact.cc b/bta/av/bta_av_aact.cc
index d646330..24703a2 100644
--- a/bta/av/bta_av_aact.cc
+++ b/bta/av/bta_av_aact.cc
@@ -2442,6 +2442,7 @@ void bta_av_str_stopped(tBTA_AV_SCB* p_scb, tBTA_AV_DATA* p_data) {
   bool sus_evt = true;
   BT_HDR* p_buf;
   uint8_t policy = HCI_ENABLE_SNIFF_MODE;
+  char *codec_name = (char *)A2DP_CodecName(p_scb->cfg.codec_info);
 
   APPL_TRACE_ERROR("%s: audio_open_cnt=%d, p_data %p", __func__,
                    bta_av_cb.audio_open_cnt, p_data);
@@ -2530,6 +2531,10 @@ void bta_av_str_stopped(tBTA_AV_SCB* p_scb, tBTA_AV_DATA* p_data) {
       (*bta_av_cb.p_cback)(BTA_AV_STOP_EVT, &bta_av_data);
     }
   }
+  if(strstr(codec_name, "aptX") != NULL) {
+    APPL_TRACE_WARNING("%s: Removing packet type limit for aptX codec stream stop", __func__);
+    AVDT_UpdateLinkPktType(p_scb->avdt_handle, btm_cb.btm_acl_pkt_types_supported);
+  }
 }
 
 /*******************************************************************************
@@ -2845,6 +2850,7 @@ void bta_av_start_ok(tBTA_AV_SCB* p_scb, tBTA_AV_DATA* p_data) {
   BT_HDR hdr;
   uint8_t policy = HCI_ENABLE_SNIFF_MODE;
   uint8_t cur_role = BTM_ROLE_UNDEFINED;
+  char *codec_name = (char *)A2DP_CodecName(p_scb->cfg.codec_info);
 
   APPL_TRACE_DEBUG("%s: wait:x%x, role:x%x", __func__, p_scb->wait,
                    p_scb->role);
@@ -2888,6 +2894,12 @@ void bta_av_start_ok(tBTA_AV_SCB* p_scb, tBTA_AV_DATA* p_data) {
     }
   }
 
+  if(strstr(codec_name, "aptX") != NULL) {
+    APPL_TRACE_WARNING("%s: Changing packet type for aptX family codec stream", __func__);
+    AVDT_UpdateLinkPktType(p_scb->avdt_handle, (btm_cb.btm_acl_pkt_types_supported |
+                HCI_PKT_TYPES_MASK_NO_3_DH1 | HCI_PKT_TYPES_MASK_NO_3_DH3 | HCI_PKT_TYPES_MASK_NO_3_DH5));
+  }
+
   if (!bta_av_link_role_ok(p_scb, A2DP_SET_ONE_BIT))
     p_scb->q_tag = BTA_AV_Q_TAG_START;
   else {
diff --git a/btif/include/btif_av.h b/btif/include/btif_av.h
index 0198d4b..da09813 100644
--- a/btif/include/btif_av.h
+++ b/btif/include/btif_av.h
@@ -39,6 +39,9 @@
 #define SOFT_HANDOFF 1
 #define RECONFIG_A2DP_PARAM 2
 
+#define APTX_HQ_LATENCY 200
+#define APTX_LL_LATENCY 70
+
 /*******************************************************************************
  *  Type definitions for callback functions
  ******************************************************************************/
diff --git a/btif/src/btif_a2dp_audio_interface.cc b/btif/src/btif_a2dp_audio_interface.cc
index f88bfc1..c0fb823 100644
--- a/btif/src/btif_a2dp_audio_interface.cc
+++ b/btif/src/btif_a2dp_audio_interface.cc
@@ -125,6 +125,7 @@ static void btif_a2dp_audio_send_mcast_status();
 static void btif_a2dp_audio_send_num_connected_devices();
 static void btif_a2dp_audio_send_connection_status();
 static void btif_a2dp_audio_send_sink_latency();
+void btif_a2dp_update_sink_latency_change();
 extern int btif_max_av_clients;
 extern bool enc_update_in_progress;
 extern tBTA_AV_HNDL btif_av_get_av_hdl_from_idx(int idx);
@@ -687,6 +688,10 @@ void btif_a2dp_audio_send_sink_latency()
   }
 }
 
+void btif_a2dp_update_sink_latency_change() {
+  btif_a2dp_audio_send_sink_latency();
+}
+
 void on_hidl_server_died() {
   LOG_INFO(LOG_TAG,"on_hidl_server_died");
   Lock lock(mtxBtAudio);
diff --git a/btif/src/btif_av.cc b/btif/src/btif_av.cc
index 9e08eaf..d94d53a 100644
--- a/btif/src/btif_av.cc
+++ b/btif/src/btif_av.cc
@@ -169,6 +169,7 @@ typedef struct {
   bool offload_state;
 #endif
   bool avdt_sync; /* for AVDT1.3 delay reporting */
+  uint16_t codec_latency;
 } btif_av_cb_t;
 
 typedef struct {
@@ -200,7 +201,7 @@ static btif_av_cb_t btif_av_cb[BTIF_AV_NUM_CB] = {
     , false, false
 #endif
     , false
-    },
+    , 0},
     { 0, {{0}}, false, 0, 0, 0, 0, std::vector<btav_a2dp_codec_config_t>(), false,
     false, false, BTIF_AV_STATE_IDLE, BTA_A2DP_SOURCE_SERVICE_ID,
     false, false, false, 0, false, false
@@ -208,7 +209,7 @@ static btif_av_cb_t btif_av_cb[BTIF_AV_NUM_CB] = {
     , false, false
 #endif
     , false
-    },
+    , 0},
 };
 
 static alarm_t* av_open_on_rc_timer = NULL;
@@ -280,6 +281,7 @@ static void btif_av_check_rc_connection_priority(void *p_data);
 static bt_status_t connect_int(RawAddress* bd_addr, uint16_t uuid);
 int btif_get_is_remote_started_idx();
 static void btif_av_reset_remote_started_flag();
+extern void btif_a2dp_update_sink_latency_change();
 
 bool isBATEnabled();
 
@@ -551,11 +553,47 @@ static void btif_update_source_codec(void* p_data) {
 
   btif_av_codec_config_req_t *req = (btif_av_codec_config_req_t *)p_data;
   btif_a2dp_source_encoder_user_config_update_req(req->codec_config, req->bd_addr);
+
+  if (req->codec_config.codec_specific_4 > 0) {
+    BTIF_TRACE_DEBUG("%s: codec_specific_4 > 0", __func__);
+    A2dpCodecConfig* current_codec = bta_av_get_a2dp_current_codec();
+    if (current_codec != nullptr) {
+      btav_a2dp_codec_config_t codec_config;
+      codec_config = current_codec->getCodecConfig();
+      BTIF_TRACE_DEBUG("%s: get codec_config", __func__);
+      if(codec_config.codec_type == BTAV_A2DP_CODEC_INDEX_SOURCE_APTX_ADAPTIVE) {
+        int index = btif_max_av_clients;
+        const uint16_t ENCODER_MODE_MASK = 0x3000;
+        const uint16_t LL_MODE_MASK = 0x2000;
+        const uint16_t HQ_MODE_MASK = 0x1000;
+        uint16_t encoder_mode = req->codec_config.codec_specific_4 & ENCODER_MODE_MASK;
+
+        if (btif_av_stream_started_ready())
+          index = btif_av_get_latest_playing_device_idx();
+        else
+          index = btif_av_get_latest_device_idx_to_start();
+
+        BTIF_TRACE_DEBUG("%s: Aptx Adaptive Codec: index = %d, encoder_mode = %d", __func__, index, encoder_mode);
+
+        if(index >= btif_max_av_clients) return;
+
+        if(encoder_mode == HQ_MODE_MASK) {
+          btif_av_cb[index].codec_latency = APTX_HQ_LATENCY;
+          btif_a2dp_update_sink_latency_change();
+        } else if (encoder_mode == LL_MODE_MASK) {
+          btif_av_cb[index].codec_latency = APTX_LL_LATENCY;
+          btif_a2dp_update_sink_latency_change();
+        }
+        BTIF_TRACE_DEBUG("%s: Aptx Adaptive codec_latency = %d", __func__, btif_av_cb[index].codec_latency);
+      }
+    }
+  }
 }
 
 static void btif_report_source_codec_state(UNUSED_ATTR void* p_data,
                                             RawAddress* bd_addr) {
   btav_a2dp_codec_config_t codec_config;
+  int index = btif_av_idx_by_bdaddr(bd_addr);
   std::vector<btav_a2dp_codec_config_t> codecs_local_capabilities;
   std::vector<btav_a2dp_codec_config_t> codecs_selectable_capabilities;
 
@@ -570,6 +608,15 @@ static void btif_report_source_codec_state(UNUSED_ATTR void* p_data,
         __func__);
     return;
   }
+
+  if (index < btif_max_av_clients) {
+    if (codec_config.codec_type == BTAV_A2DP_CODEC_INDEX_SOURCE_APTX_ADAPTIVE) {
+      if(btif_av_cb[index].codec_latency == 0)
+        btif_av_cb[index].codec_latency = APTX_HQ_LATENCY;
+    } else {
+      btif_av_cb[index].codec_latency = 0;
+    }
+  }
   if (bt_av_src_callbacks != NULL) {
     BTIF_TRACE_DEBUG("%s codec config changed BDA:0x%02X%02X%02X%02X%02X%02X", __func__,
                    bd_addr->address[0], bd_addr->address[1], bd_addr->address[2],
@@ -688,6 +735,7 @@ static bool btif_av_state_idle_handler(btif_sm_event_t event, void* p_data, int
       btif_av_cb[index].tws_device = false;
       btif_av_cb[index].offload_state = false;
 #endif
+      btif_av_cb[index].codec_latency = 0;
       for (int i = 0; i < btif_max_av_clients; i++)
         btif_av_cb[i].dual_handoff = false;
       osi_property_get("persist.vendor.service.bt.a2dp.sink", a2dp_role, "false");
@@ -4232,6 +4280,7 @@ tBTA_AV_LATENCY btif_av_get_sink_latency() {
       sink_latency = btif_av_cb[index].sink_latency;
     else
       sink_latency = BTIF_AV_DEFAULT_SINK_LATENCY;
+
   } else {
     BTIF_TRACE_DEBUG("%s, multicast enabled, calculate average sink latency", __func__);
     for (i = 0; i < btif_max_av_clients; i++) {
@@ -4866,6 +4915,15 @@ void btif_av_reset_audio_delay(tBTA_AV_HNDL hndl) {
 }
 
 uint16_t btif_av_get_audio_delay(int index) {
+  A2dpCodecConfig* current_codec = bta_av_get_a2dp_current_codec();
+  if(current_codec != nullptr) {
+    if(current_codec->codecIndex() == BTAV_A2DP_CODEC_INDEX_SOURCE_APTX_ADAPTIVE) {
+      BTIF_TRACE_WARNING("%s: Updating Aptx Adaptive specific delay: %d",
+            __func__, btif_av_cb[index].codec_latency);
+      return btif_av_cb[index].codec_latency;
+    }
+  }
+
   if (index >= 0 && index < btif_max_av_clients) {
     return btif_a2dp_control_get_audio_delay(index);
   } else {
@@ -4951,4 +5009,3 @@ bool btif_device_in_sink_role() {
         return true;
     return false;
 }
-
diff --git a/stack/avdt/avdt_api.cc b/stack/avdt/avdt_api.cc
index e68ae46..1da2952 100644
--- a/stack/avdt/avdt_api.cc
+++ b/stack/avdt/avdt_api.cc
@@ -34,6 +34,8 @@
 #include "bt_target.h"
 #include "bt_types.h"
 #include "btm_api.h"
+#include "btm_int.h"
+#include "btm_int_types.h"
 #include "btu.h"
 #include "l2c_api.h"
 #include "stack/include/a2dp_codec_api.h"
@@ -1306,3 +1308,17 @@ uint8_t AVDT_SetTraceLevel(uint8_t new_level) {
 void AVDT_UpdateMaxAvClients(uint8_t max_clients) {
     avdt_scb_set_max_av_client(max_clients);
 }
+
+/*******************************************************************************
+ *
+ * Function         AVDT_UpdateLinkPktType
+ *
+ * Description      Sends command to change packet type for a link
+ *
+ * Returns          void
+ *
+ ******************************************************************************/
+void AVDT_UpdateLinkPktType(uint8_t hdl, uint16_t packet_type) {
+    tACL_CONN* p_acl_cb = btm_bda_to_acl(avdt_scb_by_hdl(hdl)->p_ccb->peer_addr, BT_TRANSPORT_BR_EDR);
+    btm_set_packet_types(p_acl_cb, packet_type);
+}
diff --git a/stack/include/avdt_api.h b/stack/include/avdt_api.h
index 0e974b3..fa522d3 100644
--- a/stack/include/avdt_api.h
+++ b/stack/include/avdt_api.h
@@ -956,6 +956,8 @@ extern void AVDT_UpdateMaxAvClients(uint8_t num_clients);
 extern uint16_t AVDT_SendReport(uint8_t handle, AVDT_REPORT_TYPE type,
                                 tAVDT_REPORT_DATA* p_data);
 
+extern void AVDT_UpdateLinkPktType(uint8_t hdl, uint16_t packet_type);
+
 /******************************************************************************
  *
  * Function         AVDT_SetTraceLevel
-- 
2.7.4

